### PR TITLE
[MSE] MSE objects should be using thread-safe refcount

### PR DESCRIPTION
**Pull request by jyavenard**
#### 3e109296984d430615fc4d881514ca9d7d0e7afa
<pre>
[MSE] MSE objects should be using thread-safe refcount
<a href="https://bugs.webkit.org/show_bug.cgi?id=265274">https://bugs.webkit.org/show_bug.cgi?id=265274</a>
<a href="https://rdar.apple.com/118734205">rdar://118734205</a>

Reviewed by Youenn Fablet.

Make MSE objects support thread-safe refcount and weakptr.
We also move common to all MediaSourcePrivate implementation to the main class.

Fly-by fix: MediaSourcePrivateAVFObjC was keeping an unused HashMap of all SourceBufferPrivate in use, we remove it.
Use dynamicDowncast where useful.

* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::nextBestMediaEngine):
(WebCore::MediaPlayer::loadWithNextMediaEngine):
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaSourcePrivate.cpp:
(WebCore::MediaSourcePrivate::MediaSourcePrivate):
(WebCore::MediaSourcePrivate::client const):
(WebCore::MediaSourcePrivate::duration const):
(WebCore::MediaSourcePrivate::buffered const):
(WebCore::MediaSourcePrivate::waitForTarget):
(WebCore::MediaSourcePrivate::seekToTime):
* Source/WebCore/platform/graphics/MediaSourcePrivate.h:
* Source/WebCore/platform/graphics/MediaSourcePrivateClient.h:
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::removedFromMediaSource):
(WebCore::SourceBufferPrivate::currentMediaTime const):
(WebCore::SourceBufferPrivate::duration const):
(WebCore::SourceBufferPrivate::updateHighestPresentationTimestamp):
(WebCore::SourceBufferPrivate::setBufferedRanges):
(WebCore::SourceBufferPrivate::reenqueSamples):
(WebCore::SourceBufferPrivate::computeSeekTime):
(WebCore::SourceBufferPrivate::clearTrackBuffers):
(WebCore::SourceBufferPrivate::provideMediaData):
(WebCore::SourceBufferPrivate::removeCodedFrames):
(WebCore::SourceBufferPrivate::removeCodedFramesInternal):
(WebCore::SourceBufferPrivate::evictCodedFrames):
(WebCore::SourceBufferPrivate::setClient):
(WebCore::SourceBufferPrivate::didReceiveInitializationSegment):
(WebCore::SourceBufferPrivate::didUpdateFormatDescriptionForTrackId):
(WebCore::SourceBufferPrivate::didReceiveSample):
(WebCore::SourceBufferPrivate::append):
(WebCore::SourceBufferPrivate::processPendingMediaSamples):
(WebCore::SourceBufferPrivate::processMediaSample):
(WebCore::SourceBufferPrivate::resetParserState):
(WebCore::SourceBufferPrivate::memoryPressure):
(WebCore::SourceBufferPrivate::setActive):
(WebCore::SourceBufferPrivate::client const):
(WebCore::SourceBufferPrivate::isAttached const): Deleted.
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
(WebCore::SourceBufferPrivate::queue):
* Source/WebCore/platform/graphics/SourceBufferPrivateClient.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.h:
(isType):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.mm:
(WebCore::MediaSourcePrivateAVFObjC::MediaSourcePrivateAVFObjC):
(WebCore::MediaSourcePrivateAVFObjC::notifyActiveSourceBuffersChanged):
(WebCore::MediaSourcePrivateAVFObjC::durationChanged):
(WebCore::MediaSourcePrivateAVFObjC::markEndOfStream):
(WebCore::MediaSourcePrivateAVFObjC::readyState const):
(WebCore::MediaSourcePrivateAVFObjC::setReadyState):
(WebCore::MediaSourcePrivateAVFObjC::currentMediaTime const):
(WebCore::MediaSourcePrivateAVFObjC::sourceBufferKeyNeeded):
(WebCore::MediaSourcePrivateAVFObjC::setSourceBufferWithSelectedVideo):
(WebCore::MediaSourcePrivateAVFObjC::failedToCreateRenderer):
(WebCore::MediaSourcePrivateAVFObjC::duration const): Deleted.
(WebCore::MediaSourcePrivateAVFObjC::buffered): Deleted.
(WebCore::MediaSourcePrivateAVFObjC::waitForTarget): Deleted.
(WebCore::MediaSourcePrivateAVFObjC::seekToTime): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h:
(isType):
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(-[WebAVSampleBufferListener invalidate]):
(-[WebAVSampleBufferListener beginObservingLayer:]):
(-[WebAVSampleBufferListener stopObservingLayer:]):
(-[WebAVSampleBufferListener beginObservingRenderer:]):
(-[WebAVSampleBufferListener stopObservingRenderer:]):
(-[WebAVSampleBufferListener observeValueForKeyPath:ofObject:change:context:]):
(WebCore::m_logIdentifier):
(WebCore::SourceBufferPrivateAVFObjC::~SourceBufferPrivateAVFObjC):
(WebCore::SourceBufferPrivateAVFObjC::setTrackChangeCallbacks):
(WebCore::SourceBufferPrivateAVFObjC::processFormatDescriptionForTrackId):
(WebCore::SourceBufferPrivateAVFObjC::didProvideContentKeyRequestInitializationDataForTrackID):
(WebCore::SourceBufferPrivateAVFObjC::appendInternal):
(WebCore::SourceBufferPrivateAVFObjC::trackDidChangeSelected):
(WebCore::SourceBufferPrivateAVFObjC::trackDidChangeEnabled):
(WebCore::SourceBufferPrivateAVFObjC::streamDataParser const):
(WebCore::SourceBufferPrivateAVFObjC::setCDMSession):
(WebCore::SourceBufferPrivateAVFObjC::layerDidReceiveError):
(WebCore::SourceBufferPrivateAVFObjC::outputObscuredDueToInsufficientExternalProtectionChanged):
(WebCore::SourceBufferPrivateAVFObjC::flushVideo):
(WebCore::SourceBufferPrivateAVFObjC::enqueueSample):
(WebCore::SourceBufferPrivateAVFObjC::enqueueSampleBuffer):
(WebCore::SourceBufferPrivateAVFObjC::timeFudgeFactor const):
(WebCore::SourceBufferPrivateAVFObjC::notifyClientWhenReadyForMoreSamples):
(WebCore::SourceBufferPrivateAVFObjC::setVideoLayer):
(WebCore::SourceBufferPrivateAVFObjC::setDecompressionSession):
(WebCore::SourceBufferPrivateAVFObjC::player const):
(WebCore::sourceBufferMap): Deleted.
(WebCore::nextMapID): Deleted.
(WebCore::SourceBufferPrivateAVFObjC::willProvideContentKeyRequestInitializationDataForTrackID): Deleted.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.cpp:
(WebCore::MediaSourcePrivateGStreamer::MediaSourcePrivateGStreamer):
(WebCore::MediaSourcePrivateGStreamer::durationChanged):
(WebCore::MediaSourcePrivateGStreamer::waitForTarget): Deleted.
(WebCore::MediaSourcePrivateGStreamer::seekToTime): Deleted.
(WebCore::MediaSourcePrivateGStreamer::duration const): Deleted.
(WebCore::MediaSourcePrivateGStreamer::buffered): Deleted.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.h:
(isType):
* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.cpp:
(WebCore::SourceBufferPrivateGStreamer::platformMaximumBufferSize const):
* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.h:
(isType):
* Source/WebCore/platform/mock/mediasource/MockMediaSourcePrivate.cpp:
(WebCore::MockMediaSourcePrivate::MockMediaSourcePrivate):
(WebCore::MockMediaSourcePrivate::duration const): Deleted.
(WebCore::MockMediaSourcePrivate::buffered): Deleted.
(WebCore::MockMediaSourcePrivate::waitForTarget): Deleted.
(WebCore::MockMediaSourcePrivate::seekToTime): Deleted.
* Source/WebCore/platform/mock/mediasource/MockMediaSourcePrivate.h:
(isType):
* Source/WebCore/platform/mock/mediasource/MockSourceBufferPrivate.cpp:
(WebCore::MockSourceBufferPrivate::mediaSourcePrivate const):
(WebCore::MockSourceBufferPrivate::readyState const):
(WebCore::MockSourceBufferPrivate::setReadyState):
(WebCore::MockSourceBufferPrivate::enqueueSample):
* Source/WebCore/platform/mock/mediasource/MockSourceBufferPrivate.h:
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:
(WebKit::MediaSourcePrivateRemote::MediaSourcePrivateRemote):
(WebKit::MediaSourcePrivateRemote::proxyWaitForTarget):
(WebKit::MediaSourcePrivateRemote::proxySeekToTime):
(WebKit::MediaSourcePrivateRemote::waitForTarget): Deleted.
(WebKit::MediaSourcePrivateRemote::seekToTime): Deleted.
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::append):
(WebKit::SourceBufferPrivateRemote::setReadyState):
(WebKit::SourceBufferPrivateRemote::setActive):
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateDidReceiveInitializationSegment):
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateHighestPresentationTimestampChanged):
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateDurationChanged):
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateDidParseSample):
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateDidDropSample):
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateDidReceiveRenderingError):
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateReportExtraMemoryCost):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:

Canonical link: <a href="https://commits.webkit.org/271111@main">https://commits.webkit.org/271111@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/9389ed03f41f6ab7e56454445ce5ea58b5ebce9b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27383 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6023 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28632 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29610 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25064 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27852 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7934 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3421 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24870 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27646 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4821 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23508 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4196 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4345 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24504 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30246 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25020 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24926 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30486 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4364 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2508 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28442 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5831 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6591 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4821 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4750 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/platform/graphics/MediaSourcePrivate.h [hunk 1]
```diff
@@ -46,14 +46,16 @@ class LegacyCDMSession;
 #endif
 
 class WEBCORE_EXPORT MediaSourcePrivate
-    : public RefCounted<MediaSourcePrivate>
-    , public CanMakeWeakPtr<MediaSourcePrivate> {
+    : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<MediaSourcePrivate> {
```
Comment (youennf) [inline Source/WebCore/platform/graphics/MediaSourcePrivate.h:49]: With this change, we no longer know where will be destroyed this private.
Looking at the members, this seems fine.
Except, if the destructor is doing something things that need to be done in a specific thread.

File: Source/WebCore/platform/graphics/SourceBufferPrivate.cpp [hunk 2]
```diff
@@ -564,10 +570,10 @@ void SourceBufferPrivate::updateTrackIds(Vector<std::pair<AtomString, AtomString
     }
 }
 
-void SourceBufferPrivate::setClient(SourceBufferPrivateClient& client)
+void SourceBufferPrivate::setClient(Ref<SourceBufferPrivateClient> client)
```
Comment (youennf) [inline Source/WebCore/platform/graphics/SourceBufferPrivate.cpp]: Why this change, this will trigger count churning for no good reason.

File: Source/WebCore/platform/graphics/SourceBufferPrivate.cpp [hunk 3]
```diff
@@ -695,26 +701,31 @@ Ref<MediaPromise> SourceBufferPrivate::append(Ref<SharedBuffer>&& buffer)
...
+    })->whenSettled(RunLoop::current(), [weakThis = ThreadSafeWeakPtr { * this }, this, abortCount = m_abortCount](auto result) mutable -> Ref<OperationPromise> {
+        RefPtr protectedThis = weakThis.get();
+        if (!protectedThis || !result)
             return OperationPromise::createAndReject(!result ? result.error() : PlatformMediaError::BufferRemoved);
         if (abortCount != m_abortCount)
             return OperationPromise::createAndResolve();
 
+        RefPtr client = this->client();
+        if (!client)
+            return OperationPromise::createAndReject(PlatformMediaError::BufferRemoved);
```
Comment (youennf) [inline Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:713]: Why doing this after the abort check, compared to the previous patch?

File: Source/WebCore/platform/graphics/SourceBufferPrivate.cpp [hunk 4]
```diff
@@ -695,26 +701,31 @@ Ref<MediaPromise> SourceBufferPrivate::append(Ref<SharedBuffer>&& buffer)
...
+
         // Resolve the changes in TrackBuffers' buffered ranges
         // into the SourceBuffer's buffered ranges
         auto trackBuffers = trackBuffersRanges();
-        client().sourceBufferPrivateTrackBuffersChanged(trackBuffers);
+        client->sourceBufferPrivateTrackBuffersChanged(trackBuffers);
 
         Vector<Ref<MediaPromise>> promises;
```
Comment (youennf) [inline Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:720]: It would be nice if Vector<Ref<MediaPromise>, 3> promises would work, this way no malloc needed.
Maybe MediaPromise::all could take an iterable.

File: Source/WebCore/platform/graphics/SourceBufferPrivate.h [hunk 5]
```diff
@@ -74,8 +76,7 @@ enum class SourceBufferAppendMode : uint8_t {
 };
 
 class SourceBufferPrivate
-    : public RefCounted<SourceBufferPrivate>
-    , public CanMakeWeakPtr<SourceBufferPrivate>
+    : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<SourceBufferPrivate>
```
Comment (youennf) [inline Source/WebCore/platform/graphics/SourceBufferPrivate.h]: It seems we need SourceBufferPrivate to being destroyed in main thread.
Maybe we should WTF::DestructionThread::MainThread for now?

File: Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm [hunk 6]
```diff
@@ -401,34 +380,36 @@ static bool sampleBufferRenderersSupportKeySession()
...
 
             if (initialized) {
                 trackDidChangeEnabled(track, enabled);
                 return;
             }
 
-            m_pendingTrackChangeTasks.append([weakThis, trackRef = Ref { track }, enabled] {
-                if (weakThis)
-                    weakThis->trackDidChangeEnabled(trackRef, enabled);
+            m_pendingTrackChangeTasks.append([weakThis = WTFMove(weakThis), trackRef = Ref { track }, enabled] {
```
Comment (youennf) [inline Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm]: Can setEnabledChangedCallback lambda be called several times?

File: Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm [hunk 7]
```diff
@@ -608,35 +583,30 @@ static bool sampleBufferRenderersSupportKeySession()
...
 
     m_abortSemaphore = Box<Semaphore>::create(0);
-    m_parser->setWillProvideContentKeyRequestInitializationDataForTrackIDCallback([weakThis = WeakPtr { *this }, abortSemaphore = m_abortSemaphore] (uint64_t trackID) mutable {
+    m_parser->setWillProvideContentKeyRequestInitializationDataForTrackIDCallback([abortSemaphore = m_abortSemaphore] (uint64_t) mutable {
         // We must call synchronously to the main thread, as the AVStreamSession must be associated
         // with the streamDataParser before the delegate method returns.
         Box<BinarySemaphore> respondedSemaphore = Box<BinarySemaphore>::create();
-        callOnMainThread([weakThis = WTFMove(weakThis), trackID, respondedSemaphore]() {
-            if (weakThis)
-                weakThis->willProvideContentKeyRequestInitializationDataForTrackID(trackID);
```
Comment (youennf) [inline Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:639]: Why are we not calling willProvideContentKeyRequestInitializationDataForTrackID anymore?

File: Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp [hunk 8]
```diff
@@ -166,30 +166,20 @@ void MediaSourcePrivateRemote::setTimeFudgeFactor(const MediaTime& fudgeFactor)
...
-}
-
-Ref<MediaPromise> MediaSourcePrivateRemote::seekToTime(const MediaTime& time)
-{
-    ASSERT_NOT_REACHED();
-    return MediaPromise::createAndReject(PlatformMediaError::LogicError);
+    if (RefPtr client = this->client())
+        client->waitForTarget(target)->whenSettled(RunLoop::current(), WTFMove(completionHandler));
+    else
+        completionHandler(makeUnexpected(PlatformMediaError::ClientDisconnected));
```
Comment (youennf) [inline Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:174]: I tend to prefer if(!client) return completionHandler style

**Review by youennf — APPROVED** [review]

File: Source/WebCore/platform/graphics/MediaSourcePrivate.h (hunk 1 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/MediaSourcePrivate.h:49]: The privates should never contain any thread-specific objects (nor DOM objects) as they are all in platforms.

File: Source/WebCore/platform/graphics/SourceBufferPrivate.cpp (hunk 4 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:720]: it probably will work.

File: Source/WebCore/platform/graphics/SourceBufferPrivate.h (hunk 5 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/SourceBufferPrivate.h]: Good point, I have a change to remove all the map of strings ; the strings are used to store the trackID which are always unsigned numbers.

File: Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm (hunk 6 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm]: good point, particularly seeing that an incorrect move is what caused failure in the other MediaServerPrivate refcounted patch

File: Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm (hunk 7 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:639]: it did nothing.
https://searchfox.org/wubkat/rev/aa809c03fded1f6f0c99d562ee397ee2f472fee9/Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm#518-524
```
void SourceBufferPrivateAVFObjC::willProvideContentKeyRequestInitializationDataForTrackID(uint64_t trackID)
{
    if (!m_mediaSource)
        return;

    UNUSED_PARAM(trackID);
}
```

File: Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp (hunk 8 above)
Comment (jyavenard) [reply Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:174]: but that requires one more line and making `client` scope longer than it needs to be.

File: Source/WebCore/platform/graphics/SourceBufferPrivate.cpp (hunk 3 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:713]: in all, it doesn't matter the order of operation.
If aborted, 

if the client has died, the promise would have been disconnected and there's no listener anymore.

And here, it will simply get rejected in the next chained promise anyway

File: Source/WebCore/platform/graphics/SourceBufferPrivate.cpp (hunk 2 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/SourceBufferPrivate.cpp]: this is an unfortunate copy of another WIP when SourceBufferPrivate is called from a different thread, and so you need to pass a Ref. I'll remove it
